### ISSUE TITLE
Can ucp_mem_map register cuda memory?

### ISSUE THREAD

**Issue by lytofd** [openucx/ucx#4411]
Can ucp_mem_map register cuda memory  ?
If it can achives, must i install nv_peer_memory？
Thx.

**Comment by bureddy** [comment]
> Can ucp_mem_map register cuda memory ?
yes

> If it can achives, must i install nv_peer_memory？
> Thx.
from API perspective, it will work irrespective of nv_peer_memory presence. it registers with IB transports and caches the registration if nv_peer_memory exist

**Comment by lytofd** [comment]
l did a experiment  aiming at registering cuda memory.
After allocating cuda memory , its value is  0x7f93b7a00000.
But when using this value to register into UCX  like this
`ucs_status_t status;
    ucp_context_h ucp_context;
    ucp_context = get_ucp_context();
    ucp_mem_map_params_t mem_map_params;
    memset(&mem_map_params, 0, sizeof(ucp_mem_map_params_t));
    mem_map_params.field_mask = UCP_MEM_MAP_PARAM_FIELD_ADDRESS |UCP_MEM_MAP_PARAM_FIELD_LENGTH |UCP_MEM_MAP_PARAM_FIELD_FLAGS;
    mem_map_params.address = *p_address;
    mem_map_params.length = size;
    mem_map_params.flags = UCP_MEM_MAP_NONBLOCK;
    if(is_allocate)
    {
        mem_map_params.flags |= UCP_MEM_MAP_ALLOCATE;
    }
    printf("map address:%p\n",*p_address);
    ucp_mem_h memh = NULL;
    status = ucp_mem_map(ucp_context, &mem_map_params, &memh);`

parameter   *p_address= 0x7f93b7a00000    size=100, is_allocated=0
But  errors occures as belows:
-----------------------------------------------------------------------------------------------------
map address:0x7f93b7a00000
[1573561187.527490] [gpu2:151497:1]         ucp_mm.c:335  UCX  DEBUG registering user memory at 0x7f93b7a00000 length 100 mem_type host
[1573561187.527521] [gpu2:151497:1]         ucp_mm.c:129  UCX  TRACE registered address 0x7f93b7a00000 length 100 on md[2] memh[0]=0xdeadbeef
[1573561187.527531] [gpu2:151497:1]         ucp_mm.c:129  UCX  TRACE registered address 0x7f93b7a00000 length 100 on md[3] memh[1]=0xdeadbeef
[1573561187.528607] [gpu2:151497:1]          ib_md.c:692  UCX  DEBUG registered memory 0x7f93b7a00000..0x7f93b7a00070 on mlx5_0 lkey 0xdb743 rkey 0xdb743 access 0x400000000000 flags 0xe1
[1573561187.528684] [gpu2:151497:1]          ib_md.c:587  UCX  TRACE 0x7f93b7a00000..0x7f93b7a01000: changing numa policy from 0 to 1, nodemask[0]=0x3
[1573561187.528702] [gpu2:151497:1]         rcache.c:569  UCX  TRACE mlx5_0: created region 0x7f93dc845f60 [0x7f93b7a00000..0x7f93b7a00070] gt rw ref 2 lkey 0xdb743 rkey 0xdb743 atomic_rkey 0xffffffff
[1573561187.528707] [gpu2:151497:1]         ucp_mm.c:129  UCX  TRACE registered address 0x7f93b7a00000 length 100 on md[5] memh[2]=0x7f93dc845f90
[1573561187.528719] [gpu2:151497:1]         ucp_mm.c:129  UCX  TRACE registered address 0x7f93b7a00000 length 100 on md[7] memh[3]=0xdeadbeef
[1573561187.528749] [gpu2:151497:1]        knem_md.c:126  UCX  ERROR KNEM create region failed: Bad address
[1573561187.528766] [gpu2:151497:1]         rcache.c:553  UCX  DEBUG failed to register region 0x7f93dc846450 [0x7f93b7a00000..0x7f93b7a00080]: Input/output error
[1573561187.528772] [gpu2:151497:1]         ucp_mm.c:121  UCX  ERROR failed to register address 0x7f93b7a00000 length 100 on md[8]=knem: Input/output error
ucp_mem_map failed
: Invalid argument
/home/centos/rmms/interface/src/alloc.c:234  Failed to map device memory after malloc itrkey
--------------------------------------------------------------------------------------------
It seems that  UCX detect this memory as host,not cuda.

**Comment by dmitrygx** [comment]
@lytofd do you use the latest UCX?

**Comment by abellina** [comment]
Given this comment by @yosefe (https://github.com/openucx/ucx/pull/4416#issuecomment-552519913), I don't know what the benefit is of registering cuda memory (not yet anyway). Can someone explain that?

**Comment by dmitrygx** [comment]
> Can someone explain that?

@abellina to force registering of CUDA memory during initialization - it will be cached, i.e. it will not register memory on a fast path. But UCX has RCACHE, so the benefits will be imperceptible

**Comment by lytofd** [comment]
> > Can ucp_mem_map register cuda memory ?
> 
> yes
> 
> > If it can achives, must i install nv_peer_memory？
> > Thx.
> 
> from API perspective, it will work irrespective of nv_peer_memory presence. it registers with IB transports and caches the registration if nv_peer_memory exist
[](url)


> > Can ucp_mem_map register cuda memory ?
> 
> yes
> 
> > If it can achives, must i install nv_peer_memory？
> > Thx.
> 
> from API perspective, it will work irrespective of nv_peer_memory presence. it registers with IB transports and caches the registration if nv_peer_memory exist
If ucp_mem_map can register cuda memory, it means that i can pass a cuda memory address to function ucp_mem_map, and registeration could be done?
But l have met the error  as  above.
Even if the cuda memory has been registered in server side (l have tried with ucp_mem_rereg_mds instead  of ucp_mem_map,  and it returned UCS_ OK), but client side can not using ucp_put to put msg to remote cuda memory.
According to   #4416 , yosefe said that  ucp_put/get could not be used for cuda memory today, only tag api support that,    it seems that  l can use ucp_tag_send_nb to send msg to  server , and using ucp_tag_recv_nb to receive client msg?

**Comment by lytofd** [comment]
> > Can ucp_mem_map register cuda memory ?
> 
> yes
> 
> > If it can achives, must i install nv_peer_memory？
> > Thx.
> 
> from API perspective, it will work irrespective of nv_peer_memory presence. it registers with IB transports and caches the registration if nv_peer_memory exist

-----------------------------------------------------------------------------
If ucp_mem_map can register cuda memory, it means that i can pass a cuda memory address to function ucp_mem_map, and registeration could be done?
But l have met the error as above.
Even if the cuda memory has been registered in server side (l have tried with ucp_mem_rereg_mds instead of ucp_mem_map, and it returned UCS_ OK), but client side can not using ucp_put to put msg to remote cuda memory.
According to #4416 , yosefe said that ucp_put/get could not be used for cuda memory today, only tag api support that, it seems that l can use ucp_tag_send_nb to send msg to server , and using ucp_tag_recv_nb to receive client msg?

**Comment by lytofd** [comment]
> @lytofd do you use the latest UCX?

Emm，not the lastest, but not too old, l will try it again.

**Comment by dmitrygx** [comment]
> According to #4416 , yosefe said that ucp_put/get could not be used for cuda memory today, only tag api support that, it seems that l can use ucp_tag_send_nb to send msg to server , and using ucp_tag_recv_nb to receive client msg?

@lytofd yes, right


> > @lytofd do you use the latest UCX?
> 
> Emm，not the lastest, but not too old, l will try it again.

yes, pls try latest UCX after #4400 gets merged (this PR fixes `ucp_mem_map()`)

**Comment by lytofd** [comment]
> > According to #4416 , yosefe said that ucp_put/get could not be used for cuda memory today, only tag api support that, it seems that l can use ucp_tag_send_nb to send msg to server , and using ucp_tag_recv_nb to receive client msg?
> 
> @lytofd yes, right
> 
> > > @lytofd do you use the latest UCX?
> > 
> > 
> > Emm，not the lastest, but not too old, l will try it again.
> 
> yes, pls try latest UCX after #4400 gets merged (this PR fixes `ucp_mem_map()`)

Latest version actually fixed the cuda registeration problem.
Server could register its cuda memory address and pack  rkey infomation, but  when client get remote rkey info and unpacking with  ucp_ep_rkey_unpack,error  happends as below:
-------------------------------------------------------------------------------------------
[1573646897.304744] [gpu2:177186:0]    cuda_ipc_md.c:130  UCX  ERROR cuCtxGetDevice(&this_device) is failed. ret:invalid device context
[1573646897.304766] [gpu2:177186:0]       ucp_rkey.c:266  UCX  ERROR failed to unpack remote key from remote md[6]: Input/output error
[gpu2:177186:0:177186] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x38)
==== backtrace (tid: 177186) ====
---------------------------------------------------------------------------------------
Unpack funtion works normally  until  updating the latest verison of  UCX today.

**Comment by lytofd** [comment]
> > According to #4416 , yosefe said that ucp_put/get could not be used for cuda memory today, only tag api support that, it seems that l can use ucp_tag_send_nb to send msg to server , and using ucp_tag_recv_nb to receive client msg?
> 
> @lytofd yes, right
> 
> > > @lytofd do you use the latest UCX?
> > 
> > 
> > Emm，not the lastest, but not too old, l will try it again.
> 
> yes, pls try latest UCX after #4400 gets merged (this PR fixes `ucp_mem_map()`)

Latest version actually fixed the cuda registeration problem.
Server could register its cuda memory address and pack rkey infomation, but when client get remote rkey info and unpacking with ucp_ep_rkey_unpack,error happends as below:
[1573646897.304744] [gpu2:177186:0] cuda_ipc_md.c:130 UCX ERROR cuCtxGetDevice(&this_device) is failed. ret:invalid device context
[1573646897.304766] [gpu2:177186:0] ucp_rkey.c:266 UCX ERROR failed to unpack remote key from remote md[6]: Input/output error
[gpu2:177186:0:177186] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x38)
==== backtrace (tid: 177186) ====
Unpack funtion works normally until updating the latest verison of UCX today.

**Comment by dmitrygx** [comment]
> Server could register its cuda memory address and pack rkey infomation, but when client get remote rkey info and unpacking with ucp_ep_rkey_unpack,error happends as below:

@lytofd this seems to be another issue.
could you try your example with cuda_ipc disabled, e.g. `UCX_TLS=rc_x,sm,cuda_copy`

**Comment by lytofd** [comment]
> > Server could register its cuda memory address and pack rkey infomation, but when client get remote rkey info and unpacking with ucp_ep_rkey_unpack,error happends as below:
> 
> @lytofd this seems to be another issue.
> could you try your example with cuda_ipc disabled, e.g. `UCX_TLS=rc_x,sm,cuda_copy`?

Thx

**Comment by dmitrygx** [comment]
@lytofd does it help you?

**Comment by lytofd** [comment]
> @lytofd does it help you?

it works for regsteration and  l had tried with  ucp_tag_recv_nb & ucp_tag_send_nb
Server started first  with calling  ucp_tag_recv_nb and waiting for  client  sending msg with ucp_tag_send_nb.
But  signal 11 has been catched when entering  ucp_tag_recv_nb .
-------------------------------------------------------------------------------------------
`const ucp_tag_t TAG = 0x1337;
request = ucp_tag_recv_nb(worker, buffer, length, recv_datatype, TAG, 0xffff,
                              (ucp_tag_recv_callback_t)ucs_empty_function);`
--------------------------------------------------------------------------------------------
Erroe:
`Caught signal 11 (Segmentation fault: Sent by the kernel at address (nil))
==== backtrace (tid: 172781) ====
`

bt:
`0x00007f07d62fc199 in waitpid () from /lib64/libpthread.so.0
#0  0x00007f07d62fc199 in waitpid () from /lib64/libpthread.so.0
#1  0x00007f07d6980195 in ucs_debugger_attach () at debug/debug.c:782
#2  0x00007f07d6983006 in ucs_error_freeze (message=0x7f07d699e083 "Sent by the kernel") at debug/debug.c:901
#3  ucs_handle_error (message=message@entry=0x7f07d699e083 "Sent by the kernel") at debug/debug.c:1071
#4  0x00007f07d698312e in ucs_debug_handle_error_signal (signo=signo@entry=11, cause=0x7f07d699e083 "Sent by the kernel", 
    fmt=fmt@entry=0x7f07d699e335 " at address %p") at debug/debug.c:1020
#5  0x00007f07d69833ab in ucs_error_signal_handler (signo=11, info=0x7f07d73dbeb0, context=<optimized out>) at debug/debug.c:1042
#6  <signal handler called>
#7  ucp_tag_recv_nb_inner (cb=0x405c97 <ucs_empty_function>, tag_mask=65535, tag=4919, datatype=<optimized out>, count=1024, buffer=0x7ffd4c6a9090, 
    worker=0x24941c8) at tag/tag_recv.c:190
#8  ucp_tag_recv_nb (worker=0x24941c8, buffer=0x7ffd4c6a9090, count=1024, datatype=8, tag=4919, tag_mask=65535, cb=0x405c97 <ucs_empty_function>)
    at tag/tag_recv.c:179
`

**Comment by dmitrygx** [comment]
@lytofd did you ask UCP to support TAG-matching?
i.e. have to specify `UCP_FEATURE_TAG` bit to `ucp_params::features` in order to enable TAG-matching in UCP, e.g.:
```C
ucp_context_h ucp_context;
ucp_params_t ucp_params;
ucp_config_t *config;
ucs_status_t status;
...
status = ucp_config_read(NULL, NULL, &config);
...
ucp_params.field_mask = UCP_PARAM_FIELD_FEATURES;
ucp_params.features = UCP_FEATURE_TAG;
status = ucp_init(&ucp_params, config, &ucp_context);
```
could you check it pls?
